chia/data_layer/data_store.py

```diff
@@ -1073,6 +1102,77 @@ async def delete(
 
         return new_root
 
+    async def upsert(
+        self,
+        key: bytes,
+        new_value: bytes,
+        tree_id: bytes32,
+        hint_keys_values: Optional[Dict[bytes, bytes]] = None,
+        use_optimized: bool = True,
+        status: Status = Status.PENDING,
+        root: Optional[Root] = None,
+    ) -> InsertResult:
+        async with self.db_wrapper.writer():
+            if root is None:
+                root = await self.get_tree_root(tree_id=tree_id)
+
+            if hint_keys_values is None:
+                old_node = await self.get_node_by_key(key=key, tree_id=tree_id)
+                if old_node.value == new_value:
+                    raise ValueError(f"New value matches old value in upsert operation: {key.hex()}")
+                old_node_hash = old_node.hash
+            else:
+                if bytes(key) not in hint_keys_values:
```

Comment:
`key` is already `bytes`.  I did not point out all cases of this so look around a bit.

```suggestion
                if key not in hint_keys_values:
```